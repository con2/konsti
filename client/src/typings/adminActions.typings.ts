import { Game } from 'typings/game.typings';

export const SUBMIT_UPDATE_HIDDEN = 'SUBMIT_UPDATE_HIDDEN';
export const SUBMIT_GET_SETTINGS = 'SUBMIT_GET_SETTINGS';
export const SUBMIT_SELECT_SIGNUP_TIME = 'SUBMIT_SELECT_SIGNUP_TIME';
export const SUBMIT_SET_TEST_TIME = 'SUBMIT_SET_TEST_TIME';
export const SUBMIT_TOGGLE_APP_OPEN = 'SUBMIT_TOGGLE_APP_OPEN';
export const SUBMIT_RESPONSE_MESSAGE = 'SUBMIT_RESPONSE_MESSAGE';

export interface SubmitUpdateHiddenAsync {
  type: typeof SUBMIT_UPDATE_HIDDEN;
  hiddenGames: readonly Game[];
}

export interface SubmitGetSettingsAsync {
  type: typeof SUBMIT_GET_SETTINGS;
  hiddenGames: readonly Game[];
  signupTime: string;
  appOpen: boolean;
}

export interface SubmitSignupTimeAsync {
  type: typeof SUBMIT_SELECT_SIGNUP_TIME;
  signupTime: string;
}

export interface SubmitSetTestTime {
  type: typeof SUBMIT_SET_TEST_TIME;
  testTime: string;
}

export interface SubmitToggleAppOpenAsync {
  type: typeof SUBMIT_TOGGLE_APP_OPEN;
  appOpen: boolean;
}

export interface SubmitResponseMessageAsync {
  type: typeof SUBMIT_RESPONSE_MESSAGE;
  responseMessage: string;
}

export type AdminActionTypes =
  | SubmitUpdateHiddenAsync
  | SubmitGetSettingsAsync
  | SubmitSignupTimeAsync
  | SubmitSetTestTime
  | SubmitToggleAppOpenAsync
  | SubmitResponseMessageAsync;
