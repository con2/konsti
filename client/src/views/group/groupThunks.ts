import {
  postCreateGroup,
  getGroup,
  postJoinGroup,
  postLeaveGroup,
  postCloseGroup,
} from "client/services/groupServices";
import { AppThunk } from "client/types/reduxTypes";
import {
  submitLeaveGroupAsync,
  submitUpdateGroupAsync,
  submitUpdateGroupCodeAsync,
} from "client/views/group/groupSlice";
import {
  PostCloseGroupRequest,
  PostJoinGroupRequest,
} from "shared/typings/api/groups";
import { exhaustiveSwitchGuard } from "shared/utils/exhaustiveSwitchGuard";

export enum PostCreateGroupErrorMessage {
  UNKNOWN = "group.generalGroupError",
  GROUP_EXISTS = "group.error.groupExists",
  CREATOR_HAS_SIGNED_GAMES = "group.error.creatorHasSignedGames",
}

export const submitCreateGroup = (): AppThunk<
  Promise<PostCreateGroupErrorMessage | undefined>
> => {
  return async (dispatch): Promise<PostCreateGroupErrorMessage | undefined> => {
    const createGroupResponse = await postCreateGroup();

    if (createGroupResponse.status === "error") {
      switch (createGroupResponse.errorId) {
        case "groupExists":
          return PostCreateGroupErrorMessage.GROUP_EXISTS;
        case "userHasSignedGames":
          return PostCreateGroupErrorMessage.CREATOR_HAS_SIGNED_GAMES;
        case "errorFindingUser":
          return PostCreateGroupErrorMessage.UNKNOWN;
        case "unknown":
          return PostCreateGroupErrorMessage.UNKNOWN;
        default:
          exhaustiveSwitchGuard(createGroupResponse.errorId);
      }
    }

    if (createGroupResponse.status === "success") {
      dispatch(submitGetGroup(createGroupResponse.groupCode));
      dispatch(
        submitUpdateGroupCodeAsync({
          groupCode: createGroupResponse.groupCode,
          isGroupCreator: true,
        }),
      );
    }
  };
};

export enum PostJoinGroupErrorMessage {
  INVALID_GROUP_CODE = "group.invalidGroupCode",
  GROUP_NOT_EXIST = "group.groupNotExist",
  UNKNOWN = "group.generalGroupError",
  CANNOT_JOIN_OWN_GROUP = "group.error.cannotUseOwnSerial",
  REMOVE_PREVIOUS_SIGNUPS_FAILED = "group.error.removePreviousSignupsFailed",
  USER_HAS_SIGNED_GAMES = "group.error.userHasSignedGames",
  ALREADY_IN_GROUP = "group.error.alreadyInGroup",
}

export const submitJoinGroup = (
  groupRequest: PostJoinGroupRequest,
): AppThunk<Promise<PostJoinGroupErrorMessage | undefined>> => {
  return async (dispatch): Promise<PostJoinGroupErrorMessage | undefined> => {
    const joinGroupResponse = await postJoinGroup(groupRequest);

    if (joinGroupResponse.status === "error") {
      switch (joinGroupResponse.errorId) {
        case "invalidGroupCode":
          return PostJoinGroupErrorMessage.INVALID_GROUP_CODE;
        case "groupDoesNotExist":
          return PostJoinGroupErrorMessage.GROUP_NOT_EXIST;
        case "removePreviousSignupsFailed":
          return PostJoinGroupErrorMessage.REMOVE_PREVIOUS_SIGNUPS_FAILED;
        case "userHasSignedGames":
          return PostJoinGroupErrorMessage.USER_HAS_SIGNED_GAMES;
        case "errorFindingUser":
          return PostJoinGroupErrorMessage.UNKNOWN;
        case "alreadyInGroup":
          return PostJoinGroupErrorMessage.ALREADY_IN_GROUP;
        case "unknown":
          return PostJoinGroupErrorMessage.UNKNOWN;
        default:
          exhaustiveSwitchGuard(joinGroupResponse.errorId);
      }
    }

    if (joinGroupResponse.status === "success") {
      dispatch(submitGetGroup(joinGroupResponse.groupCode));
      dispatch(
        submitUpdateGroupCodeAsync({
          groupCode: joinGroupResponse.groupCode,
          isGroupCreator: false,
        }),
      );
    }
  };
};

export enum PostLeaveGroupErrorMessage {
  UNKNOWN = "group.generalGroupError",
  FAILED_TO_LEAVE = "group.generalLeaveGroupError",
}

export const submitLeaveGroup = (): AppThunk<
  Promise<PostLeaveGroupErrorMessage | undefined>
> => {
  return async (dispatch): Promise<PostLeaveGroupErrorMessage | undefined> => {
    const leaveGroupResponse = await postLeaveGroup();

    if (leaveGroupResponse.status === "error") {
      switch (leaveGroupResponse.errorId) {
        case "failedToLeave":
          return PostLeaveGroupErrorMessage.FAILED_TO_LEAVE;
        case "unknown":
          return PostLeaveGroupErrorMessage.UNKNOWN;
        default:
          exhaustiveSwitchGuard(leaveGroupResponse.errorId);
      }
    }

    if (leaveGroupResponse.status === "success") {
      dispatch(submitLeaveGroupAsync(leaveGroupResponse.groupCode));
    }
  };
};

export enum PostCloseGroupErrorMessage {
  UNKNOWN = "group.generalGroupError",
  ONLY_CREATOR_CAN_CLOSE = "group.error.onlyCreatorCanCloseGroup",
}

export const submitCloseGroup = (
  groupRequest: PostCloseGroupRequest,
): AppThunk<Promise<PostCloseGroupErrorMessage | undefined>> => {
  return async (dispatch): Promise<PostCloseGroupErrorMessage | undefined> => {
    const leaveGroupResponse = await postCloseGroup(groupRequest);

    if (leaveGroupResponse.status === "error") {
      switch (leaveGroupResponse.errorId) {
        case "onlyCreatorCanCloseGroup":
          return PostCloseGroupErrorMessage.ONLY_CREATOR_CAN_CLOSE;
        case "unknown":
          return PostCloseGroupErrorMessage.UNKNOWN;
        default:
          exhaustiveSwitchGuard(leaveGroupResponse.errorId);
      }
    }

    if (leaveGroupResponse.status === "success") {
      dispatch(submitLeaveGroupAsync(leaveGroupResponse.groupCode));
    }
  };
};

enum GetGroupErrorMessage {
  UNKNOWN = "group.generalGroupError",
}

export const submitGetGroup = (
  groupCode: string,
): AppThunk<Promise<GetGroupErrorMessage | undefined>> => {
  return async (dispatch): Promise<GetGroupErrorMessage | undefined> => {
    const getGroupResponse = await getGroup(groupCode);

    if (getGroupResponse.status === "error") {
      switch (getGroupResponse.errorId) {
        case "unknown":
          return GetGroupErrorMessage.UNKNOWN;
        default:
          exhaustiveSwitchGuard(getGroupResponse.errorId);
      }
    }

    if (getGroupResponse.status === "success") {
      dispatch(submitUpdateGroupAsync(getGroupResponse.results));
    }
  };
};
