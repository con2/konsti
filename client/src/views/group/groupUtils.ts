export const getIsGroupCreator = (
  groupCode: string,
  serial: string
): boolean => {
  if (groupCode === serial) {
    return true;
  }
  if (groupCode === "0") {
    return true;
  }
  return false;
};
