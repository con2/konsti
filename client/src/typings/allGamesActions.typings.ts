import { Game } from 'typings/game.typings';

export const SUBMIT_GET_GAMES = 'SUBMIT_GET_GAMES';

export interface SubmitGetGamesAsync {
  type: typeof SUBMIT_GET_GAMES;
  games: readonly Game[];
}

export type AllGamesActionTypes = SubmitGetGamesAsync;
