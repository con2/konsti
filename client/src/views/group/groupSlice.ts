import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { GroupState } from "client/typings/redux.typings";
import { GroupMember } from "shared/typings/models/groups";

const initialState: GroupState = {
  groupCode: "0",
  isGroupCreator: false,
  groupMembers: [],
};

interface GroupCodeUpdate {
  groupCode: string;
  isGroupCreator: boolean;
}

const groupSlice = createSlice({
  name: "group",
  initialState,
  reducers: {
    submitUpdateGroupCodeAsync(state, action: PayloadAction<GroupCodeUpdate>) {
      return {
        ...state,
        groupCode: action.payload.groupCode,
        isGroupCreator: action.payload.isGroupCreator,
      };
    },

    submitLeaveGroupAsync(state, action: PayloadAction<string>) {
      return { ...state, groupCode: action.payload, groupMembers: [] };
    },

    submitUpdateGroupAsync(
      state,
      action: PayloadAction<readonly GroupMember[]>,
    ) {
      return {
        ...state,
        groupMembers: action.payload,
      };
    },
  },
});

export const {
  submitUpdateGroupCodeAsync,
  submitLeaveGroupAsync,
  submitUpdateGroupAsync,
} = groupSlice.actions;

export const groupReducer = groupSlice.reducer;
