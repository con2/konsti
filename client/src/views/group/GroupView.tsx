import React, { FC, ReactElement, ChangeEvent } from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector, useStore } from 'react-redux';
import styled from 'styled-components';
import {
  submitJoinGroup,
  submitCreateGroup,
  submitLeaveGroup,
} from 'views/group/groupActions';
import { GroupMembersList } from 'views/group/components/GroupMembersList';
import { sleep } from 'utils/sleep';
import { config } from 'config';
import { submitSignup } from 'views/signup/signupActions';
import { loadGroupMembers } from 'utils/loadData';
import { GroupMember } from 'typings/group.typings';

import { RootState } from 'typings/redux.typings';

export const GroupView: FC = (): ReactElement => {
  const username: string = useSelector(
    (state: RootState) => state.login.username
  );
  const serial: string = useSelector((state: RootState) => state.login.serial);
  const groupCode: string = useSelector(
    (state: RootState) => state.login.groupCode
  );
  const groupMembers: readonly GroupMember[] = useSelector(
    (state: RootState) => state.login.groupMembers
  );
  const dispatch = useDispatch();
  const { t } = useTranslation();

  const [loading, setLoading] = React.useState<boolean>(false);
  const [showCreateGroup, setShowCreateGroup] = React.useState<boolean>(false);
  const [showJoinGroup, setShowJoinGroup] = React.useState<boolean>(false);
  const [joinGroupValue, setJoinGroupValue] = React.useState<string>('');
  const [message, setMessage] = React.useState<string>('');
  const [messageStyle, setMessageStyle] = React.useState<string>('');
  const [closeGroupConfirmation, setCloseGroupConfirmation] = React.useState<
    boolean
  >(false);

  const store = useStore();

  React.useEffect(() => {
    const fetchData = async (): Promise<void> => {
      await loadGroupMembers();
    };
    fetchData();
  }, [store]);

  const openGroupForming = (): void => {
    setShowCreateGroup(true);
    setShowJoinGroup(false);
  };

  const openJoinGroup = (): void => {
    setShowJoinGroup(true);
    setShowCreateGroup(false);
  };

  const createGroup = async (): Promise<void> => {
    const groupData = {
      username: username,
      groupCode: serial,
      leader: true,
      ownSerial: serial,
    };

    try {
      await dispatch(submitCreateGroup(groupData));
    } catch (error) {
      showMessage({
        message: t('generalCreateGroupError'),
        style: 'error',
      });
      return;
    }

    showMessage({ message: t('groupCreated'), style: 'success' });
  };

  // Remove all signups
  const removeSignups = async (): Promise<void> => {
    const signupData = {
      username,
      selectedGames: [],
      signupTime: 'all',
    };

    await dispatch(submitSignup(signupData));
  };

  const joinGroup = async (): Promise<void> => {
    const groupData = {
      username: username,
      groupCode: joinGroupValue,
      leader: false,
      ownSerial: serial,
    };

    try {
      await dispatch(submitJoinGroup(groupData));
    } catch (error) {
      switch (error.code) {
        case 31:
          showMessage({
            message: t('invalidGroupCode'),
            style: 'error',
          });
          return;
        case 32:
          showMessage({
            message: t('groupNotExist'),
            style: 'error',
          });
          return;
        default:
          showMessage({
            message: t('generalCreateGroupError'),
            style: 'error',
          });
          return;
      }
    }

    showMessage({ message: t('groupJoined'), style: 'success' });
    await removeSignups();
  };

  const leaveGroup = async ({ leader }: { leader: boolean }): Promise<void> => {
    setLoading(true);

    const groupData = {
      username: username,
      groupCode: groupCode,
      leader,
      ownSerial: serial,
      leaveGroup: true,
    };

    try {
      await dispatch(submitLeaveGroup(groupData));
    } catch (error) {
      switch (error.code) {
        case 36:
          showMessage({
            message: t('groupNotEmpty'),
            style: 'error',
          });
          return;
        default:
          showMessage({
            message: t('generalLeaveGroupError'),
            style: 'error',
          });
          return;
      }
    }

    showMessage({ message: t('leftGroup'), style: 'success' });
    setShowJoinGroup(false);
    setLoading(false);
  };

  const toggleCloseGroupConfirmation = (value: boolean): void => {
    setCloseGroupConfirmation(value);
  };

  const closeGroup = async ({ leader }: { leader: boolean }): Promise<void> => {
    setLoading(true);
    const groupData = {
      username: username,
      groupCode: groupCode,
      leader,
      ownSerial: serial,
      leaveGroup: true,
      closeGroup: true,
    };

    try {
      await dispatch(submitLeaveGroup(groupData));
    } catch (error) {
      showMessage({
        message: t('generalLeaveGroupError'),
        style: 'error',
      });
    }

    showMessage({ message: t('closedGroup'), style: 'success' });
    toggleCloseGroupConfirmation(false);
    setLoading(false);
  };

  const handleJoinGroupChange = (
    event: ChangeEvent<HTMLInputElement>
  ): void => {
    setJoinGroupValue(event.target.value);
  };

  const isInGroup = (): boolean => {
    if (groupCode && groupCode !== '0') {
      return true;
    }
    return false;
  };

  const showMessage = async ({
    message,
    style,
  }: {
    message: string;
    style: string;
  }): Promise<void> => {
    setMessage(message);
    setMessageStyle(style);
    await sleep(config.MESSAGE_DELAY);
    setMessage('');
    setMessageStyle('');
  };

  const groupLeader = isGroupLeader(groupCode, serial);
  const inGroup = isInGroup();

  const joinGroupInput = (
    <div>
      <FormInput
        key='joinGroup'
        placeholder={t('enterGroupLeaderCode')}
        value={joinGroupValue}
        onChange={handleJoinGroupChange}
      />
    </div>
  );

  return (
    <div className='group-view'>
      <h2>{t('pages.group')}</h2>

      <div className='group-instructions'>
        <p>{t('groupSignupGuide')}</p>
      </div>

      {groupCode === '0' && !inGroup && (
        <>
          <StyledButton
            disabled={loading}
            className={showCreateGroup ? 'active' : ''}
            onClick={() => openGroupForming()}
          >
            {t('button.createGroup')}
          </StyledButton>

          <StyledButton
            disabled={loading}
            className={showJoinGroup ? 'active' : ''}
            onClick={() => openJoinGroup()}
          >
            {t('button.joinGroup')}
          </StyledButton>

          <GroupStatusMessage className={messageStyle}>
            {message}
          </GroupStatusMessage>

          {showCreateGroup && (
            <div>
              <p>{t('createGroupConfirmationMessage')}</p>
              <p>{t('groupLeaderWarning')}</p>
              <button
                disabled={loading}
                onClick={async () => await createGroup()}
              >
                {t('button.joinGroupConfirmation')}
              </button>
            </div>
          )}

          {showJoinGroup && (
            <div>
              <p className='bold'>{t('joiningGroupWillCancelGames')}</p>

              {joinGroupInput}
              <button
                disabled={loading}
                onClick={async () => await joinGroup()}
              >
                {t('button.joinGroup')}
              </button>
            </div>
          )}
        </>
      )}

      {groupLeader && inGroup && (
        <div className='group-info'>
          <p className='group-leader-info'>
            <span className='bold'>{t('youAreGroupLeader')}</span>.{' '}
            {t('groupLeaderInfo')}
          </p>
        </div>
      )}

      {!groupLeader && inGroup && (
        <div className='group-info'>
          <p>
            <span className='bold'>{t('youAreInGroup')}</span>.{' '}
            {t('groupMemberInfo')}
          </p>
        </div>
      )}

      {inGroup && (
        <>
          <div className='group-controls'>
            {!groupLeader && (
              <button
                disabled={loading}
                onClick={async () => await leaveGroup({ leader: groupLeader })}
              >
                {t('button.leaveGroup')}
              </button>
            )}

            {groupLeader && (
              <>
                <div>
                  <button
                    disabled={loading}
                    onClick={() => toggleCloseGroupConfirmation(true)}
                  >
                    {t('button.closeGroup')}
                  </button>

                  <GroupStatusMessage className={messageStyle}>
                    {message}
                  </GroupStatusMessage>
                </div>
                {closeGroupConfirmation && (
                  <div>
                    <p>{t('closeGroupConfirmation')}</p>
                    <button
                      disabled={loading}
                      onClick={() => toggleCloseGroupConfirmation(false)}
                    >
                      {t('button.cancel')}
                    </button>

                    <WarningButton
                      disabled={loading}
                      onClick={async () =>
                        await closeGroup({ leader: groupLeader })
                      }
                    >
                      {t('button.closeGroup')}
                    </WarningButton>
                  </div>
                )}
              </>
            )}
          </div>

          <h3>{t('groupMembers')}</h3>
          <GroupMembersList groupMembers={groupMembers} />
        </>
      )}
    </div>
  );
};

export const isGroupLeader = (groupCode: string, serial: string): boolean => {
  if (groupCode === serial) {
    return true;
  }
  if (groupCode === '0') {
    return true;
  }
  return false;
};

const GroupStatusMessage = styled.span`
  font-weight: 600;
`;

const StyledButton = styled.button`
  &.active {
    background-color: ${(props) => props.theme.buttonSelected};
    border: 1px solid ${(props) => props.theme.borderActive};
  }
`;

const WarningButton = styled.button`
  background-color: ${(props) => props.theme.warning};
  color: ${(props) => props.theme.warningButtonText};
`;

const FormInput = styled.input`
  border: 1px solid ${(props) => props.theme.borderInactive};
  color: ${(props) => props.theme.buttonText};
  height: 34px;
  padding: 0 0 0 10px;
  width: 100%;
`;
