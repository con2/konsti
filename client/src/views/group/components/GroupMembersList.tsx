import React, { FC, ReactElement } from 'react';
import { useTranslation } from 'react-i18next';
import styled from 'styled-components';
import { GroupMember } from 'typings/group.typings';

export interface Props {
  groupMembers: readonly GroupMember[];
}

export const GroupMembersList: FC<Props> = (props: Props): ReactElement => {
  const { groupMembers } = props;
  const { t } = useTranslation();

  if (!groupMembers) return <GroupMembersListContainer />;

  const membersList = groupMembers.map((member, index) => {
    const leader = member.serial === member.groupCode;
    return (
      <p key={member.username}>
        {index + 1}) {member.username}{' '}
        {leader && <span>({t('groupLeader')})</span>}
      </p>
    );
  });

  return <GroupMembersListContainer>{membersList}</GroupMembersListContainer>;
};

const GroupMembersListContainer = styled.div`
  margin: 0 0 0 14px;
`;
