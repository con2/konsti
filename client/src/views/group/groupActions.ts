import { postGroup, getGroup } from 'services/groupServices';
import { GroupData, GroupMember } from 'typings/group.typings';
import { AppThunk } from 'typings/utils.typings';
import {
  SubmitUpdateGroupCodeAsync,
  SubmitGetGroupAsync,
  SubmitLeaveGroupAsync,
  SUBMIT_UPDATE_GROUP_CODE,
  SUBMIT_LEAVE_GROUP,
  SUBMIT_UPDATE_GROUP_MEMBERS,
} from 'typings/groupActions.typings';

export const submitJoinGroup = (groupData: GroupData): AppThunk => {
  return async (dispatch): Promise<void> => {
    const joinGroupResponse = await postGroup(groupData);

    if (joinGroupResponse?.status === 'error') {
      return await Promise.reject(joinGroupResponse);
    }

    if (joinGroupResponse?.status === 'success') {
      dispatch(submitGetGroup(joinGroupResponse.groupCode));
      dispatch(submitUpdateGroupCodeAsync(joinGroupResponse.groupCode));
    }
  };
};

export const submitCreateGroup = (groupData: GroupData): AppThunk => {
  return async (dispatch): Promise<void> => {
    const createGroupResponse = await postGroup(groupData);

    if (createGroupResponse?.status === 'error') {
      return await Promise.reject(createGroupResponse);
    }

    if (createGroupResponse?.status === 'success') {
      dispatch(submitGetGroup(createGroupResponse.groupCode));
      dispatch(submitUpdateGroupCodeAsync(createGroupResponse.groupCode));
    }
  };
};

const submitUpdateGroupCodeAsync = (
  groupCode: string
): SubmitUpdateGroupCodeAsync => {
  return {
    type: SUBMIT_UPDATE_GROUP_CODE,
    groupCode,
  };
};

export const submitGetGroup = (groupCode: string): AppThunk => {
  return async (dispatch): Promise<void> => {
    const getGroupResponse = await getGroup(groupCode);

    if (getGroupResponse?.status === 'error') {
      return await Promise.reject(getGroupResponse);
    }

    if (getGroupResponse?.status === 'success') {
      dispatch(submitGetGroupAsync(getGroupResponse.results));
    }
  };
};

const submitGetGroupAsync = (
  groupMembers: readonly GroupMember[]
): SubmitGetGroupAsync => {
  return {
    type: SUBMIT_UPDATE_GROUP_MEMBERS,
    groupMembers,
  };
};

export const submitLeaveGroup = (groupData: GroupData): AppThunk => {
  return async (dispatch): Promise<void> => {
    const leaveGroupResponse = await postGroup(groupData);

    if (leaveGroupResponse?.status === 'error') {
      return await Promise.reject(leaveGroupResponse);
    }

    if (leaveGroupResponse?.status === 'success') {
      dispatch(submitLeaveGroupAsync(leaveGroupResponse.groupCode));
    }
  };
};

const submitLeaveGroupAsync = (groupCode: string): SubmitLeaveGroupAsync => {
  return {
    type: SUBMIT_LEAVE_GROUP,
    groupCode,
  };
};
