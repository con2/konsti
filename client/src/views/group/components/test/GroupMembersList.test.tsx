import React from 'react';
import { shallow } from 'enzyme';
import { GroupMembersList, Props } from '../GroupMembersList';
import { GroupMember } from 'typings/group.typings';

const groupMembers: GroupMember[] = [];

describe('GroupMembersList', () => {
  it('should render correctly', () => {
    const props: Props = { groupMembers };
    const component = shallow(<GroupMembersList {...props} />);
    expect(component).toMatchSnapshot();
  });
});
