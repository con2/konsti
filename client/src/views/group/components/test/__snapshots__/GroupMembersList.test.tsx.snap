// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupMembersList should render correctly 1`] = `<GroupMembersList__GroupMembersListContainer />`;
